geps/gep-1619.md

```diff
@@ -157,49 +281,401 @@ Generally, the implementation API programs the dataplane API; however these two
+configurations, like allowing customization of the cookie name, could prevent certain implementations from conforming to
+the spec. In other words, opting for a higher-level API provides better interoperability among our implementations.
+
+#### Target Persona
+
+Referring to the [Gateway API Security Model](https://gateway-api.sigs.k8s.io/concepts/security-model/#roles-and-personas),
+the target kubernetes role/persona for session persistence are application developers, as mentioned in the [When does an application require session persistence?](#when-does-an-application-require-session-persistence)
+section. It is the responsibility of the application developers to adjust the persistence configuration to ensure the
+functionality of their applications.
+
+#### Prior Art
+
+Referring to our [Implementations](#Implementations) table on session persistence, the majority of Gateway API
+implementations designed session persistence in their APIs to be attached to a service or backends. This should be
+considered cautiously, as making associations to Gateway API's notion of Gateway, Route, and Service to other
+implementation's objects is hard to directly translate. The idea of a route in Gateway API is often not the same as a
+route in any given implementation.
+
+#### Metaresource Policy Design
+
+In order to apply session persistence configuration to both a service and a route, we will implement it as a
+metaresource policy, as outlined in [Policy Attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment/#direct-policy-attachment).
+The metaresource is named `SessionPersistencePolicy` and is only responsible for configuring session persistence for
+services or routes. It is defined as a [Direct Policy Attachment](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/#direct-policy-attachment)
+without defaults or overrides, applied to the targeted service, HTTPRoute, or GRPCRoute.
+
+Attaching the `SessionPersistencePolicy` metaresource to a service will be a core support level feature, while attaching
+it to a route will be considered extended. Implementations must support services for conformance, while routes are
+considered optional. This distinction arises because most existing implementations primarily support attaching to a
+service, while attaching to a route is less common and involves greater complexities (see [API Attachment Point: Route vs. Service](#api-attachment-point-route-vs-service)
+for more details).
+
+#### API Attachment Point: Route vs. Service
+
+The `SessionPersistencePolicy` metaresource can target both a service and route object. The core functionality of
+attaching it to either is generally identical. The distinction lies in where the configuration gets propagated: when
+attached to a route, it can define session persistence for multiple services, whereas attaching it to a service defines
```

Comment:
> Attaching to an entire HttpRoute would be trickier.

I thought attaching to an entire HTTPRoute is easier and only to a section (e.g. route-rule) is trickier in the current API? Unless you are talking about something else

Reply:
What is the advantage of supporting both types? It seems that specifying `SessionPersistencePolicy` in HTTPRoute would provide more granular control and could eliminate the need for supporting attachment to `Service`

Reply:
Not getting into 'eliminate the need' - since we agreed to support both modes. 

My point was that attaching to HttpRoute will mean that ALL backends used in routes will be persistent. 
So if you have a HttpRoute with /foo -> fooBackend, /bar -> barBackend - both backends will have persistence. 
Maybe that's ok - I don't fully understand how sessions work with routes, my focus has been on the Service attachment. 

I'm not asking for any change - just wanted to confirm attaching to the entire route is the intent here, for my own understanding.

Reply:
> Not getting into 'eliminate the need' - since we agreed to support both modes.

@costinm can you please share more context here.

> My point was that attaching to HttpRoute will mean that ALL backends used in routes will be persistent.
So if you have a HttpRoute with /foo -> fooBackend, /bar -> barBackend - both backends will have persistence.
Maybe that's ok - I don't fully understand how sessions work with routes, my focus has been on the Service attachment.
This might not provide sufficient granularity. for example when a single K8s service port is used for various gRPC services.

Reply:
I am not entirely following the rationale behind the 'attachment to xRoute' model instead of per [route matcher](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.HTTPRouteRule). Imagine a case where the same K8s Service handles multiple gRPC Services (which leads to different request paths) with different session persistency requirements.  It seems that with the current model users of this API would need to define multiple separate xRoutes with redundant/repetitive configuration (e.g. hostnames, backendrefs, traffic splitting, etc)

Reply:
@va7ish yes your use-case is valid and should be considered. I can tell you that there are very high level proposals in the works to allow attaching metaresources to route-rules (what you call route matcher) and they would come later.  In the interest of time and getting this SSP proposal approved and merged we would like to defer your suggestion to these later proposals so we can have SSPs that can be attached to routes. Also like you said there is the work-around of splitting a route into multiple routes to provide granularity

Reply:
> Does anyone have an issue with dropping entire route matching (xRoute) in favor for sectional route matching (of which we will probably add in another PR of this GEP)? @sanjaypujare?

If I understood it correctly: switch to the model where the session persistence policy is only attachable to a specific xRoute Rule (not the entire xRoute) then I think that's good and gets my support.

Reply:
I think it's still useful to be able to target a policy to the entire route (is there any advantage in dropping it?), but if we are able to support targeting a section (i.e. a route rule) then it's okay to drop it but I am in favor of keeping it. So one can apply a policy to a 1) service 2) route rule or 3) entire route.

Reply:
We discussed this in the [first meeting](https://docs.google.com/document/d/1DYYyxM-O_JXuLPyWijabUbcl1cRBSFpX-OFi6eSijs8/edit#heading=h.gb5npubb0ftd) we had and we decided:
- Attaching to routes is important
- Attaching to services is important
- Attaching to route `matcher` section (when available) is important

So I've updated the GEP to reflect this, added a paragraph reference to anticipating route section targeting support.